File: Llama3.CSharp/Transformer.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace Llama3.CSharp;

public class Weights
{
    public Tensor TokenEmbd { get; set; }
    public Tensor[] RmsAttW { get; set; }
    public Tensor[] Wq { get; set; }
    public Tensor[] Wk { get; set; }
    public Tensor[] Wv { get; set; }
    public Tensor[] Wo { get; set; }
    public Tensor[] RmsFfnW { get; set; }
    public Tensor[] W1 { get; set; }
    public Tensor[] W2 { get; set; }
    public Tensor[] W3 { get; set; }
    public Tensor RmsFinalW { get; set; }
    public Tensor OutputW { get; set; }

    public static Weights Load(Config config, GGUF gguf, Memory<byte> mmapData)
    {
        var w = new Weights
        {
            TokenEmbd = GetTensor(gguf, mmapData, "token_embd.weight"),
            RmsAttW = new Tensor[config.NLayers],
            Wq = new Tensor[config.NLayers],
            Wk = new Tensor[config.NLayers],
            Wv = new Tensor[config.NLayers],
            Wo = new Tensor[config.NLayers],
            RmsFfnW = new Tensor[config.NLayers],
            W1 = new Tensor[config.NLayers],
            W2 = new Tensor[config.NLayers],
            W3 = new Tensor[config.NLayers]
        };

        for (int i = 0; i < config.NLayers; i++)
        {
            w.RmsAttW[i] = GetTensor(gguf, mmapData, $"blk.{i}.attn_norm.weight");
            w.Wq[i] = GetTensor(gguf, mmapData, $"blk.{i}.attn_q.weight");
            w.Wk[i] = GetTensor(gguf, mmapData, $"blk.{i}.attn_k.weight");
            w.Wv[i] = GetTensor(gguf, mmapData, $"blk.{i}.attn_v.weight");
            w.Wo[i] = GetTensor(gguf, mmapData, $"blk.{i}.attn_output.weight");
            w.RmsFfnW[i] = GetTensor(gguf, mmapData, $"blk.{i}.ffn_norm.weight");
            w.W1[i] = GetTensor(gguf, mmapData, $"blk.{i}.ffn_gate.weight");
            w.W2[i] = GetTensor(gguf, mmapData, $"blk.{i}.ffn_down.weight");
            w.W3[i] = GetTensor(gguf, mmapData, $"blk.{i}.ffn_up.weight");
        }

        w.RmsFinalW = GetTensor(gguf, mmapData, "output_norm.weight");
        w.OutputW = gguf.TensorInfos.ContainsKey("output.weight")
            ? GetTensor(gguf, mmapData, "output.weight")
            : w.TokenEmbd;

        return w;
    }

    private static Tensor GetTensor(GGUF gguf, Memory<byte> mmapData, string name)
    {
        if (!gguf.TensorInfos.TryGetValue(name, out var info)) throw new Exception($"Tensor not found: {name}");
        long offset = gguf.TensorDataOffset + info.Offset;
        long size = info.Type.ByteSizeFor(info.NumElements());
        return new Tensor
        {
            Data = mmapData.Slice((int)offset, (int)size),
            Type = info.Type,
            NumElements = info.NumElements()
        };
    }
}

public class State
{
    public float[] X { get; }
    public float[] Xb { get; }
    public float[] Xb2 { get; }
    public float[] Hb { get; }
    public float[] Hb2 { get; }
    public float[] Q { get; }
    public float[] K { get; }
    public float[] V { get; }
    public float[] Att { get; }
    public float[] Logits { get; }
    public float[] KeyCache { get; }
    public float[] ValueCache { get; }

    public State(Config config)
    {
        int kvDim = config.NKvHeads * config.HeadSize;
        X = new float[config.Dim];
        Xb = new float[config.Dim];
        Xb2 = new float[config.Dim];
        Hb = new float[config.HiddenDim];
        Hb2 = new float[config.HiddenDim];
        Q = new float[config.Dim];
        K = new float[config.Dim];
        V = new float[config.Dim];
        Att = new float[config.NHeads * config.SeqLen];
        Logits = new float[config.VocabSize];
        KeyCache = new float[config.NLayers * config.SeqLen * kvDim];
        ValueCache = new float[config.NLayers * config.SeqLen * kvDim];
    }
}

public static class Transformer
{
    public static void Forward(Config config, State state, Weights weights, int token, int pos)
    {
        int dim = config.Dim;
        int hdim = config.HiddenDim;
        int hsz = config.HeadSize;
        int nhead = config.NHeads;
        int nkv = config.NKvHeads;
        int kvDim = nkv * hsz;
        int kvMul = nhead / nkv;

        float[] embd = weights.TokenEmbd.CopyToF32(token * dim, dim);
        Array.Copy(embd, state.X, dim);

        for (int l = 0; l < config.NLayers; l++)
        {
            MathKernels.RmsNorm(state.Xb, state.X, weights.RmsAttW[l].CopyToF32(0, dim), config.Eps);

            MathKernels.MatMul(state.Q, state.Xb, weights.Wq[l], nhead * hsz, dim);
            MathKernels.MatMul(state.K, state.Xb, weights.Wk[l], nkv * hsz, dim);
            MathKernels.MatMul(state.V, state.Xb, weights.Wv[l], nkv * hsz, dim);

            // RoPE
            for (int h = 0; h < nhead; h++)
            {
                for (int i = 0; i < hsz / 2; i++)
                {
                    float freq = 1.0f / (float)Math.Pow(config.RopeTheta, (float)(2 * i) / hsz);
                    if (config.RopeScaling)
                    {
                        float wlen = 2.0f * (float)Math.PI / freq;
                        float low = config.OldContextLen / config.LoFreqFactor;
                        float high = config.OldContextLen / config.HiFreqFactor;
                        if (wlen > low) freq /= config.ScaleFactor;
                        else if (wlen > high)
                        {
                            float smooth = (config.OldContextLen / wlen - config.LoFreqFactor) / (config.HiFreqFactor - config.LoFreqFactor);
                            freq = (1.0f - smooth) * freq / config.ScaleFactor + smooth * freq;
                        }
                    }
                    float val = pos * freq;
                    float fcr = (float)Math.Cos(val);
                    float fci = (float)Math.Sin(val);

                    float q0 = state.Q[h * hsz + i * 2];
                    float q1 = state.Q[h * hsz + i * 2 + 1];
                    state.Q[h * hsz + i * 2] = q0 * fcr - q1 * fci;
                    state.Q[h * hsz + i * 2 + 1] = q0 * fci + q1 * fcr;

                    if (h < nkv)
                    {
                        float k0 = state.K[h * hsz + i * 2];
                        float k1 = state.K[h * hsz + i * 2 + 1];
                        state.K[h * hsz + i * 2] = k0 * fcr - k1 * fci;
                        state.K[h * hsz + i * 2 + 1] = k0 * fci + k1 * fcr;
                    }
                }
            }

            int loff = l * config.SeqLen * kvDim;
            Array.Copy(state.K, 0, state.KeyCache, loff + pos * kvDim, kvDim);
            Array.Copy(state.V, 0, state.ValueCache, loff + pos * kvDim, kvDim);

            // Multi-head Attention
            for (int h = 0; h < nhead; h++)
            {
                int qOffset = h * hsz;
                int attOffset = h * config.SeqLen;
                for (int t = 0; t <= pos; t++)
                {
                    int kOffset = loff + t * kvDim + (h / kvMul) * hsz;
                    float score = 0;
                    for (int i = 0; i < hsz; i++) score += state.Q[qOffset + i] * state.KeyCache[kOffset + i];
                    score /= (float)Math.Sqrt(hsz);
                    state.Att[attOffset + t] = score;
                }

                MathKernels.Softmax(state.Att.AsSpan(attOffset, pos + 1));

                int xbOffset = h * hsz;
                for (int i = 0; i < hsz; i++) state.Xb[xbOffset + i] = 0;
                for (int t = 0; t <= pos; t++)
                {
                    int vOffset = loff + t * kvDim + (h / kvMul) * hsz;
                    float a = state.Att[attOffset + t];
                    for (int i = 0; i < hsz; i++) state.Xb[xbOffset + i] += a * state.ValueCache[vOffset + i];
                }
            }

            MathKernels.MatMul(state.Xb2, state.Xb, weights.Wo[l], dim, dim);
            for (int i = 0; i < dim; i++) state.X[i] += state.Xb2[i];

            MathKernels.RmsNorm(state.Xb, state.X, weights.RmsFfnW[l].CopyToF32(0, dim), config.Eps);
            MathKernels.MatMul(state.Hb, state.Xb, weights.W1[l], hdim, dim);
            MathKernels.MatMul(state.Hb2, state.Xb, weights.W3[l], hdim, dim);

            for (int i = 0; i < hdim; i++)
            {
                float val = state.Hb[i];
                val *= 1.0f / (1.0f + (float)Math.Exp(-val));
                state.Hb[i] = val * state.Hb2[i];
            }

            MathKernels.MatMul(state.Xb, state.Hb, weights.W2[l], dim, hdim);
            for (int i = 0; i < dim; i++) state.X[i] += state.Xb[i];
        }

        MathKernels.RmsNorm(state.X, state.X, weights.RmsFinalW.CopyToF32(0, dim), config.Eps);
        MathKernels.MatMul(state.Logits, state.X, weights.OutputW, config.VocabSize, dim);
    }
}

File: Llama3.Rust/src/transformer.rs
use crate::gguf::{GGUF, GGMLType};
use crate::config::Config;
use crate::tensors::{Tensor, rmsnorm, matmul, dot_f32};
use crate::tokenizer::Tokenizer;
use crate::sampler::Sampler;
use std::time::Instant;

pub struct Weights<'a> {
    pub token_embd: Tensor<'a>,
    pub rms_att_w: Vec<Tensor<'a>>,
    pub wq: Vec<Tensor<'a>>,
    pub wk: Vec<Tensor<'a>>,
    pub wv: Vec<Tensor<'a>>,
    pub wo: Vec<Tensor<'a>>,
    pub rms_ffn_w: Vec<Tensor<'a>>,
    pub w1: Vec<Tensor<'a>>,
    pub w2: Vec<Tensor<'a>>,
    pub w3: Vec<Tensor<'a>>,
    pub rms_final_w: Tensor<'a>,
    pub output_w: Tensor<'a>,
}

impl<'a> Weights<'a> {
    pub fn load(config: &Config, gguf: &'a GGUF, mmap_data: &'a [u8]) -> Self {
        let mut rms_att_w = Vec::with_capacity(config.n_layers);
        let mut wq = Vec::with_capacity(config.n_layers);
        let mut wk = Vec::with_capacity(config.n_layers);
        let mut wv = Vec::with_capacity(config.n_layers);
        let mut wo = Vec::with_capacity(config.n_layers);
        let mut rms_ffn_w = Vec::with_capacity(config.n_layers);
        let mut w1 = Vec::with_capacity(config.n_layers);
        let mut w2 = Vec::with_capacity(config.n_layers);
        let mut w3 = Vec::with_capacity(config.n_layers);

        for i in 0..config.n_layers {
            rms_att_w.push(get_tensor(gguf, mmap_data, &format!("blk.{}.attn_norm.weight", i)));
            wq.push(get_tensor(gguf, mmap_data, &format!("blk.{}.attn_q.weight", i)));
            wk.push(get_tensor(gguf, mmap_data, &format!("blk.{}.attn_k.weight", i)));
            wv.push(get_tensor(gguf, mmap_data, &format!("blk.{}.attn_v.weight", i)));
            wo.push(get_tensor(gguf, mmap_data, &format!("blk.{}.attn_output.weight", i)));
            rms_ffn_w.push(get_tensor(gguf, mmap_data, &format!("blk.{}.ffn_norm.weight", i)));
            w1.push(get_tensor(gguf, mmap_data, &format!("blk.{}.ffn_gate.weight", i)));
            w2.push(get_tensor(gguf, mmap_data, &format!("blk.{}.ffn_down.weight", i)));
            w3.push(get_tensor(gguf, mmap_data, &format!("blk.{}.ffn_up.weight", i)));
        }

        let token_embd = get_tensor(gguf, mmap_data, "token_embd.weight");
        let rms_final_w = get_tensor(gguf, mmap_data, "output_norm.weight");
        let output_w = if gguf.tensor_infos.contains_key("output.weight") {
            get_tensor(gguf, mmap_data, "output.weight")
        } else {
            // Tie embeddings
            Tensor { data: token_embd.data, ggml_type: token_embd.ggml_type, num_elements: token_embd.num_elements }
        };

        Weights {
            token_embd, rms_att_w, wq, wk, wv, wo, rms_ffn_w, w1, w2, w3, rms_final_w, output_w
        }
    }
}

fn get_tensor<'a>(gguf: &'a GGUF, mmap_data: &'a [u8], name: &str) -> Tensor<'a> {
    let info = gguf.tensor_infos.get(name).expect(&format!("Tensor not found: {}", name));
    let offset = gguf.tensor_data_offset + info.offset;
    let size = info.ggml_type.byte_size_for(info.num_elements());
    Tensor {
        data: &mmap_data[offset as usize..(offset as usize + size)],
        ggml_type: info.ggml_type,
        num_elements: info.num_elements(),
    }
}

pub struct State {
    pub x: Vec<f32>,
    pub xb: Vec<f32>,
    pub xb2: Vec<f32>,
    pub hb: Vec<f32>,
    pub hb2: Vec<f32>,
    pub q: Vec<f32>,
    pub k: Vec<f32>,
    pub v: Vec<f32>,
    pub att: Vec<f32>,
    pub logits: Vec<f32>,
    pub key_cache: Vec<f32>,
    pub value_cache: Vec<f32>,
}

impl State {
    pub fn new(config: &Config) -> Self {
        let kv_dim = config.n_kv_heads * config.head_size;
        State {
            x: vec![0.0; config.dim],
            xb: vec![0.0; config.dim],
            xb2: vec![0.0; config.dim],
            hb: vec![0.0; config.hidden_dim],
            hb2: vec![0.0; config.hidden_dim],
            q: vec![0.0; config.dim],
            k: vec![0.0; config.dim],
            v: vec![0.0; config.dim],
            att: vec![0.0; config.n_heads * config.seq_len],
            logits: vec![0.0; config.vocab_size],
            key_cache: vec![0.0; config.n_layers * config.seq_len * kv_dim],
            value_cache: vec![0.0; config.n_layers * config.seq_len * kv_dim],
        }
    }
}

pub fn forward(config: &Config, state: &mut State, weights: &Weights, token: u32, pos: usize) {
    let dim = config.dim;
    let hdim = config.hidden_dim;
    let hsz = config.head_size;
    let nhead = config.n_heads;
    let nkv = config.n_kv_heads;
    let kv_dim = nkv * hsz;
    let kv_mul = nhead / nkv;

    weights.token_embd.copy_to_f32(token as usize * dim, dim, &mut state.x);

    let mut weight_f32 = vec![0.0; dim.max(hdim)];

    for l in 0..config.n_layers {
        weights.rms_att_w[l].copy_to_f32(0, dim, &mut weight_f32[..dim]);
        rmsnorm(&mut state.xb, &state.x, &weight_f32[..dim], config.eps);

        matmul(&mut state.q, &state.xb, &weights.wq[l], nhead * hsz, dim);
        matmul(&mut state.k, &state.xb, &weights.wk[l], nkv * hsz, dim);
        matmul(&mut state.v, &state.xb, &weights.wv[l], nkv * hsz, dim);

        // RoPE
        for h in 0..nhead {
            for i in 0..hsz / 2 {
                let mut freq = 1.0 / (config.rope_theta.powf((2 * i) as f32 / hsz as f32));
                if config.rope_scaling {
                    let wlen = 2.0 * std::f32::consts::PI / freq;
                    let low = config.old_context_len / config.lo_freq_factor;
                    let high = config.old_context_len / config.hi_freq_factor;
                    if wlen > low { freq /= config.scale_factor; }
                    else if wlen > high {
                        let smooth = (config.old_context_len / wlen - config.lo_freq_factor) / (config.hi_freq_factor - config.lo_freq_factor);
                        freq = (1.0 - smooth) * freq / config.scale_factor + smooth * freq;
                    }
                }
                let val = pos as f32 * freq;
                let fcr = val.cos();
                let fci = val.sin();

                let q0 = state.q[h * hsz + i * 2];
                let q1 = state.q[h * hsz + i * 2 + 1];
                state.q[h * hsz + i * 2] = q0 * fcr - q1 * fci;
                state.q[h * hsz + i * 2 + 1] = q0 * fci + q1 * fcr;

                if h < nkv {
                    let k0 = state.k[h * hsz + i * 2];
                    let k1 = state.k[h * hsz + i * 2 + 1];
                    state.k[h * hsz + i * 2] = k0 * fcr - k1 * fci;
                    state.k[h * hsz + i * 2 + 1] = k0 * fci + k1 * fcr;
                }
            }
        }

        let loff = l * config.seq_len * kv_dim;
        state.key_cache[loff + pos * kv_dim..loff + (pos + 1) * kv_dim].copy_from_slice(&state.k[0..kv_dim]);
        state.value_cache[loff + pos * kv_dim..loff + (pos + 1) * kv_dim].copy_from_slice(&state.v[0..kv_dim]);

        // Multi-head Attention
        for h in 0..nhead {
            let q_offset = h * hsz;
            let att_offset = h * config.seq_len;
            for t in 0..=pos {
                let k_offset = loff + t * kv_dim + (h / kv_mul) * hsz;
                let score = dot_f32(&state.q[q_offset..q_offset + hsz], &state.key_cache[k_offset..k_offset + hsz]);
                state.att[att_offset + t] = score / (hsz as f32).sqrt();
            }

            crate::tensors::softmax(&mut state.att[att_offset..att_offset + pos + 1]);

            let xb_offset = h * hsz;
            for i in 0..hsz { state.xb[xb_offset + i] = 0.0; }
            for t in 0..=pos {
                let v_offset = loff + t * kv_dim + (h / kv_mul) * hsz;
                let a = state.att[att_offset + t];
                for i in 0..hsz {
                    state.xb[xb_offset + i] += a * state.value_cache[v_offset + i];
                }
            }
        }

        matmul(&mut state.xb2, &state.xb, &weights.wo[l], dim, dim);
        for i in 0..dim { state.x[i] += state.xb2[i]; }

        weights.rms_ffn_w[l].copy_to_f32(0, dim, &mut weight_f32[..dim]);
        rmsnorm(&mut state.xb, &state.x, &weight_f32[..dim], config.eps);

        matmul(&mut state.hb, &state.xb, &weights.w1[l], hdim, dim);
        matmul(&mut state.hb2, &state.xb, &weights.w3[l], hdim, dim);

        for i in 0..hdim {
            let val = state.hb[i];
            state.hb[i] = (val / (1.0 + (-val).exp())) * state.hb2[i];
        }

        matmul(&mut state.xb, &state.hb, &weights.w2[l], dim, hdim);
        for i in 0..dim { state.x[i] += state.xb[i]; }
    }

    weights.rms_final_w.copy_to_f32(0, dim, &mut weight_f32[..dim]);
    let x_clone = state.x.clone();
    rmsnorm(&mut state.x, &x_clone, &weight_f32[..dim], config.eps);

    matmul(&mut state.logits, &state.x, &weights.output_w, config.vocab_size, dim);
}
